### DISCUSSION TITLE
Training issues

### DISCUSSION THREAD

**Discussion by bridgetsiek** [talmolab/sleap#1838] [Help!]
Hi, 

I have been getting the message "Instances were predicted on 0 frames". I troubleshoot it a couple of times by changing the parameters,  but I am still not able to get it to predict frames. I tried the ways the other posts mentioned like changing the max stride and filters etc. 
![image](https://github.com/talmolab/sleap/assets/108977368/a106ccb8-a386-4a86-a655-13ddd8c630e2)

![image](https://github.com/talmolab/sleap/assets/108977368/d1de2ec6-ea64-44e6-9879-f00554794137)
I hope to get some help and explain why it doesn't work, I am new to programming and SLEAP. 
Thank you so much!

**Comment by talmo** [comment]
Hi @bridgetsiek,

Hmm, nothing sticks out to me in the configuration, so there might be something else going on here.

Do you mind sharing a (few) screenshot(s) of your video(s) with your labels? This will help us get a sense of what kind of landmarks you want to track and we can go from there.

Thanks!

Talmo

**Comment by bridgetsiek** [comment]
Hi Talmo,
Thank you for responding! I have attached 4 screenshots of the video with labels. 
![image](https://github.com/talmolab/sleap/assets/108977368/cafd666c-57f4-4567-8e23-2c2abbcbe419)
![image](https://github.com/talmolab/sleap/assets/108977368/55047adb-e46d-48f3-b8fc-bc1e9c2b1850)
![image](https://github.com/talmolab/sleap/assets/108977368/2c5535bb-dc6b-431e-b986-18fea643bfaf)
![image](https://github.com/talmolab/sleap/assets/108977368/91da6f15-34dd-4b55-b374-3d4287f5fdf8)

Thank you,
Bridget

**Comment by talmo** [comment]
Oo yeah, I see, that's going to be a bit tricky since side views are inherently tougher due to the drastic differences in appearance when you have the rat in different orientations.

It should definitely still be doable, but we often have to train with quite a bit more labels for side views.

One other thing to try is one of the pretrained encoder backbones instead of the UNet since those tend to have decent visual feature detectors for side views. EfficientNetB0 is a good one to start with, or one of the ResNets. Regardless though, I'd definitely expect more labeling to be required.

Some more tips:

- It might not be a bad idea to start off with a simpler skeleton (at least without the tail) just to get a training configuration dialed in that works well on your data. You can go back and add more nodes to the skeleton later (definitely save it to a new project via File --> Save as...).
- I think you're already likely noticing it, but it looks like you have some motion blur which will make it particularly challenging to precisely annotate particular body parts. If it's hard for you to be consistent about it, then the models will have a hard time with it as well. The solution here is to decrease the exposure time on your camera (<5 ms is ideal).
- You might consider doing distal/proximal instead of left/right for the lateralized nodes. This might make it easier for the models to generalize to when the rat is facing either side on the wheel.

Give these a go and let us know how it goes!

Talmo

**Reply by bridgetsiek** [reply]
Hi Talmo, 
Thank you so much! I used efficientnetb0 and it was able to predict instances on 10 frames. However, it saved the predicted instances on a separate .slp file. Do I need to merge it back to the original project to correct the predictions and run training again, or should I leave it in the separate .slp file and train using this new file? 

Thank you again!

Bridget

**Reply by talmo** [reply]
Hi @bridgetsiek,

Great! SLEAP will save the predictions to a separate file but will also import them into the current project. If you're not seeing them in the current project, it's possible that the predictions were empty (i.e., the model failed to optimize appropriately).

You could check by opening the separate predictions file in a new window and seeing if there are any predicted instances.

Cheers,

Talmo

**Reply by bridgetsiek** [reply]
Hi Talmo,

There are predictions in the new file but not in the current project.

Thank you,
Bridget

**Reply by talmo** [reply]
Ok, are you able to import and merge those in (from the file menu) and see the predictions in your project?

**Comment by dana-c-hckl** [comment]
Hey everybody,

I have a problem running training with Sleap too. The error code is: 

2024-08-07 14:44:27.390602: W tensorflow/core/kernels/data/cache_dataset_ops.cc:768] The calling iterator did not fully read the dataset being cached. In order to avoid unexpected truncation of the dataset, the partially cached contents of the dataset  will be discarded. This can happen if you have an input pipeline similar to `dataset.cache().take(k).repeat()`. You should use `dataset.take(k).cache().repeat()` instead.
INFO:sleap.nn.callbacks:Closing the reporter controller/context.
INFO:sleap.nn.callbacks:Closing the training controller socket/context.
Run Path: C:/Users/Engert Lab/Desktop\models\bottom_up240807_144417.single_instance.n=90


Is their any experience someone can share? I already doublechecked tensorflow and GPU. 

Looking forward.

**Reply by talmo** [reply]
Hi @dana-c-hckl,

Apologies, we missed this reply last week!

The logs you pasted don't see to indicate any error, just regular warnings that are no cause for concern.

Do you mind [creating a new discussion](https://github.com/talmolab/sleap/discussions/new?category=help) and sharing a bit more about what kind of errors you're getting, what your project looks like (a screenshot of the GUI is fine), and a longer set of logs beyond the bit you pasted here?

Thanks!

Talmo
